subgraphs/uniswap-v3/src/common/price/price.ts

```diff
@@ -139,46 +144,105 @@ export function findUSDPricePerToken(
         }
-        if (pool.inputTokens[1] == token.id) {
-          const token0 = getOrCreateToken(event, pool.inputTokens[0], false);
-          // get the derived whitelist in pool
-          const whitelistTokenValueLocked =
-            poolAmounts.inputTokenBalances[0].times(token0.lastPriceUSD!);
-          if (
-            whitelistTokenValueLocked.gt(largestWhitelistTokenValue) &&
-            whitelistTokenValueLocked.gt(
-              NetworkConfigs.getMinimumLiquidityThreshold()
-            )
-          ) {
-            largestWhitelistTokenValue = whitelistTokenValueLocked;
-            // token0 per our token * whitelist token per token0
-            priceSoFar = poolAmounts.tokenPrices[0].times(
-              token0.lastPriceUSD as BigDecimal
-            );
+
+        const whitelistToken = getOrCreateToken(
+          event,
+          pool.inputTokens[whitelistTokenIndex],
+          false
+        );
+        // get the derived whitelist token in pool
+        const whitelistTokenValueLocked = poolAmounts.inputTokenBalances[
+          whitelistTokenIndex
+        ].times(whitelistToken.lastPriceUSD!);
+
+        // Check if it meets criteria to update prices.
+        if (
+          whitelistTokenValueLocked.gt(largestWhitelistTokenValue) &&
+          whitelistTokenValueLocked.gt(
+            NetworkConfigs.getMinimumLiquidityThreshold()
+          )
+        ) {
+          // Calculate new price of a token and TVL of token in this pool.
+          const newPriceSoFar = poolAmounts.tokenPrices[
+            whitelistTokenIndex
+          ].times(whitelistToken.lastPriceUSD as BigDecimal);
+
+          const newTokenTotalValueLocked = convertTokenToDecimal(
+            token._totalSupply,
+            token.decimals
+          ).times(newPriceSoFar);
+
+          // If price is too high, skip this pool
+          if (newTokenTotalValueLocked.gt(BIGDECIMAL_TEN_BILLION)) {
+            log.warning("Price too high for token: {} from pool: {}", [
+              token.id.toHexString(),
+              pool.id.toHexString(),
+            ]);
+            continue;
           }
+
+          // Set new price and largest pool for pricing.
+          largestWhitelistTokenValue = whitelistTokenValueLocked;
+          priceSoFar = newPriceSoFar;
         }
       }
     }
   }
+
+  // Buffer token pricings that would cause large spikes on the protocol level
+  const protocol = getOrCreateProtocol();
+  const tokenTVLDelta = absBigDecimal(
+    priceSoFar
+      .times(convertTokenToDecimal(token._totalSupply, token.decimals))
+      .minus(token._totalValueLockedUSD)
+  );
+  const protocolTVLPercentageDelta = absBigDecimal(
+    safeDiv(tokenTVLDelta, protocol.totalValueLockedUSD)
+  );
+  if (protocolTVLPercentageDelta.gt(BIGDECIMAL_FIVE_PERCENT)) {
+    log.warning("Price too high for token: {} from pool: {}", [
+      token.id.toHexString(),
+      priceSoFar.toString(),
+    ]);
+    token._largeTVLImpactBuffer += 1;
```

Comment:
I like this, should we also add a count for the number of times a price is skipped bc it is over 10bb?

Reply:
I decided not to do that, because there is never an instance where a token TVL has been close, and if it does go over, its most certainly wrong.